### PR TITLE
Only call Free in unmanaged->managed stubs when ownership has been transfered to the callee

### LINKED ISSUE #85795: ComInterfaceGenerator generates cleanup stage for stateless marshalled parameters when it shouldn't
Parameters marshalled by stateless marshallers are being freed in too many scenarios. They should only be freed in unmanaged to managed ref scenarios (on the pointer that was passed in, not the value that is returned at the end of the method. In and Out parameters should not be cleaned up. Stateful marshallers can have free called on them as they can know which MarshalMode they are in and decide to actually free the underlying memory or not.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #85795

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/AttributedMarshallingModelGeneratorFactory.cs [hunk 1]
```diff
@@ -245,7 +245,12 @@ private IMarshallingGenerator CreateCustomNativeTypeMarshaller(TypePositionInfo
                     marshallingStrategy = new StatelessCallerAllocatedBufferMarshalling(marshallingStrategy, marshallerData.MarshallerType.Syntax, marshallerData.BufferElementType.Syntax, isLinearCollectionMarshalling: false);
 
                 if (marshallerData.Shape.HasFlag(MarshallerShape.Free))
-                    marshallingStrategy = new StatelessFreeMarshalling(marshallingStrategy, marshallerData.MarshallerType.Syntax);
+                {
+                    if (context.Direction == MarshalDirection.ManagedToUnmanaged || !context.AdditionalTemporaryStateLivesAcrossStages)
+                        marshallingStrategy = new StatelessFreeMarshalling(marshallingStrategy, marshallerData.MarshallerType.Syntax);
+                    else if (info.RefKind == RefKind.Ref)
+                        marshallingStrategy = new StatelessByRefFreeMarshalling(marshallingStrategy, marshallerData.MarshallerType.Syntax);
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/AttributedMarshallingModelGeneratorFactory.cs:252]:
```suggestion
                    if (context.Direction == MarshalDirection.ManagedToUnmanaged || !context.AdditionalTemporaryStateLivesAcrossStages)
                    {
                        marshallingStrategy = new StatelessFreeMarshalling(marshallingStrategy, marshallerData.MarshallerType.Syntax);
                    }
                    else if (info.RefKind == RefKind.Ref)
                    {
                        marshallingStrategy = new StatelessByRefFreeMarshalling(marshallingStrategy, marshallerData.MarshallerType.Syntax);
                    }
```

File: src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Tests/UnmanagedToManagedCustomMarshallingTests.cs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                    }
+                }
+
+                [VirtualMethodIndex(0, ImplicitThisParameter = true)]
+                [return: MarshalUsing(typeof(IntWrapperMarshallerToIntWithFreeCounts))]
+                IntWrapper GetData();
+                [VirtualMethodIndex(1, ImplicitThisParameter = true)]
+                void SetData([MarshalUsing(typeof(IntWrapperMarshallerToIntWithFreeCounts))] IntWrapper x);
+                [VirtualMethodIndex(2, ImplicitThisParameter = true)]
+                void ExchangeData([MarshalUsing(typeof(IntWrapperMarshallerToIntWithFreeCounts))] ref IntWrapper data);
```
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Tests/UnmanagedToManagedCustomMarshallingTests.cs:46]: Do we also want a case for `out`?

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs [hunk 3]
```diff
@@ -295,6 +295,93 @@ public IEnumerable<StatementSyntax> GenerateCleanupStatements(TypePositionInfo i
...
+        public IEnumerable<StatementSyntax> GenerateUnmarshalCaptureStatements(TypePositionInfo info, StubCodeContext context)
+        {
+            foreach (StatementSyntax statement in _innerMarshaller.GenerateUnmarshalCaptureStatements(info, context))
+            {
+                yield return statement;
+            }
+
+            // Now that we've captured the new value to pass to the caller, we need to make sure that we free the old one.
+
+            // <freeUnmanaged> = true;
```
Comment (jtschuster) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/StatelessMarshallingStrategy.cs:374]: What scenarios would we use this StatelessByRefFreeMarshalling type but not call GenerateUnmarshalCaptureStatements? Do we/could we have a test for one of those scenarios? In the generated code I can only find code that sets `freeUnmanaged` to false, then immediately to true again.

**Comment by jkoritzinsky** [comment]
I've updated this PR to correctly handle lifetimes for collection marshallers (stateless and stateful). I've opened #86608 to track additional work to complete this story.

**Review by jtschuster — APPROVED** [review]
LGTM

**Comment by jkoritzinsky** [comment]
All failures are known

**Pull request merged**